File: Add Code Here/cpp/heapSort.cpp
#include "iostream"
using namespace std;


void heapify(int arr[], int n, int i){
    int largest = i;
    int leftChild = 2*i+1;
    int rightChild = 2*i+2;

    if(leftChild < n && arr[largest] < arr[leftChild])
        largest = leftChild;

    if(rightChild < n && arr[largest] < arr[rightChild])
        largest = rightChild;

    if(largest != i){
        swap(arr[largest], arr[i]);
        heapify(arr, n, largest);
    }

}


void heapSort(int arr[], int n){
    if(n <= 1) return;

    int size = n-1;
    while(size > 0){

        swap(arr[0], arr[size]);
        size--;

        heapify(arr, size+1, 0);
    }

}


int main(){

    int arr[9] = {20,10,8,0,-5,50, 43, 7,8};
    int n=sizeof(arr)/sizeof(arr[0]);

    for(int i=n/2; i>=0; i--){
        heapify(arr, n, i);
    }


    for(int i=0; i<n; i++){
        cout << arr[i] << " ";
    }
    cout << endl;

    heapSort(arr, n);

    for(int i=0; i<n; i++){
        cout << arr[i] << " ";
    }
    cout << endl;


    return 0;
}

File: Add Code Here/PYTHON/Heap_sort.py
def heapify(nums, heap_size, root_index):

    largest = root_index
    left_child = (2 * root_index) + 1
    right_child = (2 * root_index) + 2


    if left_child < heap_size and nums[left_child] > nums[largest]:
        largest = left_child


    if right_child < heap_size and nums[right_child] > nums[largest]:
        largest = right_child


    if largest != root_index:
        nums[root_index], nums[largest] = nums[largest], nums[root_index]

        heapify(nums, heap_size, largest)


def heap_sort(nums):
    n = len(nums)
    for i in range(n, -1, -1):
        heapify(nums, n, i)


    for i in range(n - 1, 0, -1):
        nums[i], nums[0] = nums[0], nums[i]
        heapify(nums, i, 0)



random_list_of_nums = [35, 12, 43, 8, 51]
heap_sort(random_list_of_nums)
print(random_list_of_nums)
